cloud/storage/core/libs/common/dynamic_persistent_table.h

```diff
@@ -0,0 +1,631 @@
+        GapSpaceSize += DescriptorsPtr[index].DataSize;
+
+        if (index + 1 == NextFreeRecordIndex) {
+            NextFreeRecordIndex--;
+            HeaderPtr->NextFreeRecordIndex = NextFreeRecordIndex;
+        } else {
+            FreeRecordIndexes.push_back(index);
+        }
+        return true;
+    }
+
+    ui64 CountRecords() const
+    {
+        return NextFreeRecordIndex - FreeRecordIndexes.size();
+    }
+
+    void Clear()
+    {
+        NextFreeRecordIndex = 0;
+        NextDataOffset = 0;
+        GapSpaceSize = 0;
+        HeadDataIndex = InvalidIndex;
+        TailDataIndex = InvalidIndex;
+        FreeRecordIndexes.clear();
+
+        FileMap->ResizeAndRemap(0, 0);
+        FileMap.reset();
+        Init();
+    }
+
+    TMemoryOutput GetAllocatedRecordDataWriter(ui64 index)
+    {
+        if (index >= MaxRecords ||
+            DescriptorsPtr[index].State != ERecordState::Allocated)
+        {
+            return TMemoryOutput(nullptr, 0);
+        }
+
+        return TMemoryOutput(
+            DataPtr + DescriptorsPtr[index].DataOffset,
+            DescriptorsPtr[index].DataSize);
+    }
+
+    bool WriteRecordData(ui64 index, const void* data, ui64 size)
+    {
+        if (index >= MaxRecords || size > DescriptorsPtr[index].DataSize ||
+            data == nullptr ||
+            DescriptorsPtr[index].State == ERecordState::Free)
+        {
+            return false;
+        }
+
+        char* recordData = DataPtr + DescriptorsPtr[index].DataOffset;
+        std::memcpy(recordData, data, size);
+        if (size < DescriptorsPtr[index].DataSize) {
+            GapSpaceSize += DescriptorsPtr[index].DataSize - size;
+            DescriptorsPtr[index].DataSize = size;
+        }
+
+        DescriptorsPtr[index].CRC32 =
+            Crc32c(recordData, DescriptorsPtr[index].DataSize);
+
+        return true;
+    }
+
+private:
+    void AddToDataList(ui64 index)
+    {
+        DescriptorsPtr[index].PrevDataIndex = InvalidIndex;
+        DescriptorsPtr[index].NextDataIndex = InvalidIndex;
+
+        if (HeadDataIndex == InvalidIndex) {
+            HeadDataIndex = index;
+            TailDataIndex = index;
+        } else {
+            DescriptorsPtr[TailDataIndex].NextDataIndex = index;
+            DescriptorsPtr[index].PrevDataIndex = TailDataIndex;
+            TailDataIndex = index;
+        }
+    }
+
+    void RemoveFromDataList(ui64 index)
+    {
+        ui64 prevIndex = DescriptorsPtr[index].PrevDataIndex;
+        ui64 nextIndex = DescriptorsPtr[index].NextDataIndex;
+
+        if (prevIndex != InvalidIndex) {
+            DescriptorsPtr[prevIndex].NextDataIndex = nextIndex;
+        } else {
+            HeadDataIndex = nextIndex;
+        }
+
+        if (nextIndex != InvalidIndex) {
+            DescriptorsPtr[nextIndex].PrevDataIndex = prevIndex;
+        } else {
+            TailDataIndex = prevIndex;
+        }
+
+        DescriptorsPtr[index].PrevDataIndex = InvalidIndex;
```

Comment:
The implementation of updating `PrevDataIndex` and `NextDataIndex` doesn't look atomic and crash consistent.
For example when calling `DeleteRecord` you will call `RemoveFromDataList` first and if you crush after setting `PrevDataIndex` to `InvalidIndex` your doubly linked list becomes corrupted so in the next call to DeleteRecord with same index the function will incorrectly set `HeadDataIndex` to `DescriptorsPtr[index].NextDataIndex` effectively discarding some data entries.
You need some journaling mechanism similar to what I've implemented in the `PrepareCompactRecord`